File: Telerik/FirstControl/04.ExpressionAgain.java
package Telerik;

import java.util.*;

public class ExpressionAgain {
    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        char[] digits = scan.nextLine().toCharArray();
        int controlExp = Integer.parseInt(scan.nextLine());
        int counter = 0;
        Queue<String> expressions = new LinkedList<String>();
        expressions.add(String.valueOf(digits[0]));
        for (int i = 1; i < digits.length; i++) {
            int tempCount = expressions.size();
            for (int j = 0; j < tempCount; j++) {
                String exp = expressions.remove();
                expressions.add(exp + "*" + digits[i]);
                expressions.add(exp + "+" + digits[i]);
                expressions.add(exp + "-" + digits[i]);
                if (digits[i - 1] != '0') {
                    expressions.add(exp + "" + digits[i]);
                } else if (exp.length() >= 2 && exp.charAt(exp.length() - 2) != '*' && exp.charAt(exp.length() - 2) != '+' && exp.charAt(exp.length() - 2) != '-') {
                    expressions.add(exp + "" + digits[i]);
                }
            }
        }
        while (expressions.size() > 0) {
            String expression = expressions.remove();
            if (EvaluateExression(expression) == controlExp) {
                counter++;
            }
            //System.out.println(expression);
            //System.out.println(EvaluateExression(expression));
        }
        System.out.println(counter);
    }

    public static int EvaluateExression(String expression) {
        int sum = 0;
        String[] matchesNumbers = expression.split("[+*-]{1}");
        String[] matchesOperators = expression.split("\\d+");
        matchesNumbers = clean(matchesNumbers);
        matchesOperators = clean(matchesOperators);
        sum = Integer.parseInt(matchesNumbers[0]);
        for (int i = 1; i < matchesNumbers.length; i++) {
            switch (matchesOperators[i - 1]) {
                case "+":
                    sum += Integer.parseInt(matchesNumbers[i]);
                    break;
                case "-":
                    sum -= Integer.parseInt(matchesNumbers[i]);
                    break;
                case "*":
                    sum *= Integer.parseInt(matchesNumbers[i]);
                    break;
            }
        }
        return sum;
    }

    public static String[] clean(final String[] arr) {
        List<String> list = new ArrayList<String>(Arrays.asList(arr));
        list.removeAll(Collections.singleton(null));
        list.removeAll(Collections.singleton(""));
        return list.toArray(new String[list.size()]);
    }
}

File: Telerik/FirstControl/04.ExpressionAgain.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

class ExpressionAgain
{
    static void Main()
    {
        char[] digits = Console.ReadLine().ToCharArray();
        int controlExp = int.Parse(Console.ReadLine());
        int counter = 0;
        Queue<string> expressions = new Queue<string>();        
        expressions.Enqueue(digits[0].ToString());
        for (int i = 1; i < digits.Length; i++)
        {
            int tempCount = expressions.Count;
            for (int j = 0; j < tempCount; j++)
            {
                string exp = expressions.Dequeue();
                expressions.Enqueue(exp + "*" + digits[i].ToString());
                expressions.Enqueue(exp + "+" + digits[i].ToString());
                expressions.Enqueue(exp + "-" + digits[i].ToString());
                if (digits[i - 1] != '0')
                {
                    expressions.Enqueue(exp + "" + digits[i].ToString());
                }
                else if (exp.Length >= 2 && exp[exp.Length - 2] != '*' && exp[exp.Length - 2] != '+' && exp[exp.Length - 2] != '-')
                {
                    expressions.Enqueue(exp + "" + digits[i].ToString());
                }                
            }
        }
        while (expressions.Count > 0)
        {
            string expression = expressions.Dequeue();
            if (EvaluateExression(expression) == controlExp)
            {
                counter++;
            }
            //Console.WriteLine(expression);
            //Console.WriteLine(EvaluateExression(expression));
        }
        Console.WriteLine(counter);
    }

    static int EvaluateExression(string expression)
    {
        int sum = 0;
        string patternNumbers = @"\d+";
        string patternOperators = @"[+*-]{1}";
        MatchCollection matchesNumbers = Regex.Matches(expression, patternNumbers);
        MatchCollection matchesOperators = Regex.Matches(expression, patternOperators);
        sum = int.Parse(matchesNumbers[0].Value);
        for (int i = 1; i < matchesNumbers.Count; i++)
        {
            switch (matchesOperators[i - 1].Value)
            {
                case "+":
                    sum += int.Parse(matchesNumbers[i].Value);
                    break;
                case "-":
                    sum -= int.Parse(matchesNumbers[i].Value);
                    break;
                case "*":
                    sum *= int.Parse(matchesNumbers[i].Value);
                    break; 
            }
        }
        return sum;
    }
}

